File: algorithms/2_medium/F/02924_find_champion_ii/solution.c
#define MAX_ALLOWED 100

int findChampion(const int n, int** edges, const int len, const int* cols) {
    if (n < 1 || n > MAX_ALLOWED) {
        goto invalid_input;
    }

    int8_t parent[MAX_ALLOWED];
    for (int i = 0; i < n; i += 1) { parent[i] = -1; }

    for (int i = 0; i < len; i += 1) {
        const int u = edges[i][0];
        const int v = edges[i][1];

        if (u < 0 || v < 0 || u >= n || v >= n) {
            goto invalid_input;
        }

        parent[v] = (int8_t)u;
    }

    int res = -1;
    for (int i = 0; i < n; i += 1) {
        if (parent[i] == -1) {
            if (res == -1) {
                res = i;
            } else {
                return -1;
            }
        }
    }

    return res;

invalid_input:
    fprintf(stderr, "invalid input\n");
    exit(1);
}

File: algorithms/2_medium/F/02924_find_champion_ii/solution.rs
impl Solution {
    const MAX_ALLOWED: usize = 100;

    pub fn find_champion(n: i32, edges: Vec<Vec<i32>>) -> i32 {
        assert!(n >= 1 && (n as usize) <= Self::MAX_ALLOWED);
        let mut parent = [-1_i8; Self::MAX_ALLOWED];

        for e in edges {
            assert!(e.len() == 2);
            let (u, v) = (e[0], e[1]);
            assert!(u >= 0 && v >= 0 && u < n && v < n);

            parent[v as usize] = u as i8;
        }

        let mut res = -1_i32;
        for i in 0..n {
            if parent[i as usize] == -1 {
                if res == -1 {
                    res = i as i32;
                } else {
                    return -1;
                }
            }
        }

        res
    }
}
